File: sop-java-picocli/src/main/java/sop/cli/picocli/commands/ExtractCertCmd.java
// SPDX-FileCopyrightText: 2021 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.cli.picocli.commands;

import java.io.IOException;

import picocli.CommandLine;
import sop.Ready;
import sop.cli.picocli.SopCLI;
import sop.exception.SOPGPException;
import sop.operation.ExtractCert;

@CommandLine.Command(name = "extract-cert",
        resourceBundle = "msg_extract-cert",
        exitCodeOnInvalidInput = SOPGPException.UnsupportedOption.EXIT_CODE)
public class ExtractCertCmd extends AbstractSopCmd {

    @CommandLine.Option(names = "--no-armor",
            negatable = true)
    boolean armor = true;

    @Override
    public void run() {
        ExtractCert extractCert = throwIfUnsupportedSubcommand(
                SopCLI.getSop().extractCert(), "extract-cert");

        if (!armor) {
            extractCert.noArmor();
        }

        try {
            Ready ready = extractCert.key(System.in);
            ready.writeTo(System.out);
        } catch (IOException e) {
            throw new RuntimeException(e);
        } catch (SOPGPException.BadData badData) {
            String errorMsg = getMsg("sop.error.input.stdin_not_a_private_key");
            throw new SOPGPException.BadData(errorMsg, badData);
        }
    }
}

File: sop-java-picocli/src/main/kotlin/sop/cli/picocli/commands/ExtractCertCmd.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.cli.picocli.commands

import java.io.IOException
import picocli.CommandLine.Command
import picocli.CommandLine.Option
import sop.cli.picocli.SopCLI
import sop.exception.SOPGPException
import sop.exception.SOPGPException.BadData

@Command(
    name = "extract-cert",
    resourceBundle = "msg_extract-cert",
    exitCodeOnInvalidInput = SOPGPException.UnsupportedOption.EXIT_CODE)
class ExtractCertCmd : AbstractSopCmd() {

    @Option(names = [OPT_NO_ARMOR], negatable = true) var armor = true

    override fun run() {
        val extractCert =
            throwIfUnsupportedSubcommand(SopCLI.getSop().extractCert(), "extract-cert")

        if (!armor) {
            throwIfUnsupportedOption(OPT_NO_ARMOR) { extractCert.noArmor() }
        }

        try {
            val ready = extractCert.key(System.`in`)
            ready.writeTo(System.out)
        } catch (e: IOException) {
            throw RuntimeException(e)
        } catch (badData: BadData) {
            val errorMsg = getMsg("sop.error.input.stdin_not_a_private_key")
            throw BadData(errorMsg, badData)
        }
    }

    companion object {
        const val OPT_NO_ARMOR = "--no-armor"
    }
}
